### PR TITLE
Speed up `is project excluded`

### PR DESCRIPTION
**Pull request by fviernau**
See individual commits for the details

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1557)

### CODE REVIEW

File: model/src/main/kotlin/config/Excludes.kt [hunk 1]
```diff
@@ -73,42 +73,11 @@ data class Excludes(
...
-        ortResult.analyzer?.result?.projects?.all { project ->
-            isProjectExcluded(project, ortResult) || project.scopes.all { scope ->
-                isScopeExcluded(scope, project, ortResult) || !scope.contains(id)
-            }
-        } ?: false
-
     /**
      * True if any [path exclude][paths] matches [path].
      */
     fun isPathExcluded(path: String) = paths.any { it.matches(path) }
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/Excludes.kt]: It would be more consistent to also move this function and `isScopeExcluded()` as part of the refactoring.

File: model/src/main/kotlin/OrtResult.kt [hunk 2]
```diff
@@ -76,6 +77,18 @@ data class OrtResult(
     @JsonInclude(JsonInclude.Include.NON_EMPTY)
     val data: CustomData = mutableMapOf()
 
+    @JsonIgnore
+    private val excludedProjects: Set<Identifier> =
+        getProjects()
+            .filter { project ->
+                val projectExclude = getExcludes().findProjectExclude(project, this)
+                val pathExcludes = getExcludes().findPathExcludes(project, this)
+                val isExcluded = projectExclude != null || pathExcludes.isNotEmpty()
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/OrtResult.kt]: Don't you also need to check if `projectExclude.isWholeProjectExcluded == true`?

File: model/src/main/kotlin/OrtResult.kt [hunk 3]
```diff
@@ -196,7 +191,7 @@ data class OrtResult(
 
                         findings.getOrPut(result.id) { mutableMapOf() }[finding] =
                                 // Only add matching excludes if all license locations are excluded.
-                            if (isExcluded) matchingExcludes.toList() else emptyList()
+                                if (isExcluded) matchingExcludes.toList() else emptyList()
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/OrtResult.kt]: Unrelated change.

File: model/src/main/kotlin/OrtResult.kt [hunk 4]
```diff
@@ -311,7 +305,33 @@ data class OrtResult(
...
+    fun isExcluded(id: Identifier): Boolean =
+        getProjects().find { it.id == id }?.let {
+            // An excluded project could still be included as a dependency of another non-excluded project.
+            isProjectExcluded(it) && isPackageExcluded(id)
+        } ?: isPackageExcluded(id)
+
+
+    /**
+     * True if all occurrences of the package identified by [id] in the [ortResult] are excluded by this [Excludes]
+     * configuration.
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/OrtResult.kt]: I would remove "by this [Excludes] configuration", it doesn't add valuable information and it's confusing what "this" refers to. Or if you want to mention it say something like "by the [Excludes] defined in the [RepositoryConfiguration] of [repository]."

File: model/src/main/kotlin/OrtResult.kt [hunk 5]
```diff
@@ -311,7 +305,33 @@ data class OrtResult(
...
+     */
+    fun isPackageExcluded(id: Identifier): Boolean =
+        getProjects().all { project ->
+            isProjectExcluded(project) || project.scopes.all { scope ->
+                getExcludes().isScopeExcluded(scope, project, this) || !scope.contains(id)
+            }
+        }
+
+    /**
+     * True if the [project] is excluded by this [Excludes] configuration.
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/OrtResult.kt]: I would remove "by this [Excludes] configuration", it doesn't add valuable information and it's confusing what "this" refers to. Or if you want to mention it say something like "by the [Excludes] defined in the [RepositoryConfiguration] of [repository]."

File: model/src/main/kotlin/OrtResult.kt [hunk 6]
```diff
@@ -326,12 +340,8 @@ data class OrtResult(
...
      * True if the [project] is excluded by this [Excludes] configuration.
      */
-    fun isProjectExcluded(project: Project): Boolean =
-        getDefinitionFilePathRelativeToAnalyzerRoot(project).let { definitionFilePath ->
-            val excludes = getExcludes()
-            excludes.projects.any { it.matches(definitionFilePath) && it.isWholeProjectExcluded }
-                    || excludes.paths.any { it.matches(definitionFilePath) }
-        }
+    fun isProjectExcluded(project: Project): Boolean
+            = excludedProjects.contains(project.id)
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/OrtResult.kt]: Could be a one-liner.

File: model/src/main/kotlin/OrtResult.kt [hunk 7]
```diff
@@ -311,7 +319,27 @@ data class OrtResult(
      * Return true if the project or package with the given identifier is excluded.
      */
     @Suppress("UNUSED") // This is intended to be mostly used via scripting.
-    fun isExcluded(id: Identifier): Boolean = getExcludes().isExcluded(id, this)
+    fun isExcluded(id: Identifier): Boolean =
+        getProjects().find { it.id == id }?.let {
+            // An excluded project could still be included as a dependency of another non-excluded project.
+            isProjectExcluded(it) && isPackageExcluded(id)
+        } ?: isPackageExcluded(id)
+
```
Comment (sschuberth) [inline model/src/main/kotlin/OrtResult.kt]: Nit: Consecutive blank lines. Maybe you can fix this retrospectively, @fviernau?

File: model/src/test/kotlin/config/ExcludesTest.kt [hunk 8]
```diff
@@ -28,6 +28,7 @@ import com.here.ort.model.PackageReference
 import com.here.ort.model.Project
 import com.here.ort.model.Repository
 import com.here.ort.model.Scope
+import io.kotlintest.TestCase
```
Comment (sschuberth) [inline model/src/test/kotlin/config/ExcludesTest.kt]: Nit: This should be moved to the import block below. Maybe you can fix this retrospectively, @fviernau?

File: model/src/test/kotlin/config/ExcludesTest.kt [hunk 9]
```diff
@@ -41,6 +42,30 @@ import io.kotlintest.shouldBe
 import io.kotlintest.specs.WordSpec
 
 class ExcludesTest : WordSpec() {
+
```
Comment (sschuberth) [inline model/src/test/kotlin/config/ExcludesTest.kt]: Nit: Superfluous blank line. Maybe you can fix this retrospectively, @fviernau?
